editors/code/src/util.ts

```diff
@@ -82,3 +86,49 @@ export function isRustDocument(document: vscode.TextDocument): document is RustD
 export function isRustEditor(editor: vscode.TextEditor): editor is RustEditor {
     return isRustDocument(editor.document);
 }
+
+export function createWorkspaceWithNewLocation(workspace: vscode.WorkspaceFolder, newLoc: vscode.Uri) {
+    return {
+        ...workspace,
+        name: path.basename(newLoc.fsPath),
+        uri: newLoc,
+    };
+}
+
+// searches up the folder structure until it finds a Cargo.toml
+export async function nearestParentWithCargoToml(
```

Comment:
We already have this implemented in rust (`find_cargo_toml` in ra_project_model), and I think we should strive for keeping as much as possible in rust, as it is sharable between the editors then.

Reply:
@matklad 
> Took a look. I must say this feels like a pretty big change to how everything works, which departs a lot from what is already implemented/what I envision, so it was a bit hard to review for me :)

Not sure what it takes away from what is already implemented / what visions it's breaking?
Is it that anything is done from the client side?

> The big piece is that we should strive to keep almost everything in Rust. I think the only thing here which genuinely belongs to the editor is didChangeWorkspaceFolders event, for everything else, I don't see a fundamental reason why it can't be done at the editor's side.

That's is a good point, I saw this an very minor change to improve QoL for vscode, but I can see how it might be better to do 100% server side for many other reasons. It took me a couple of hours to code up (mostly reading vscode issues). If that's what is ruining this pr for you I could redo it in the server, however there it will actually affect all users of it (but maybe that is what we want?).

>  still don't fully understand what specific use-cases this covers, but here's my ideas about what we can do to make handing multi-projects better

There are many uses cases

1. You have a large project where rust is deep subfolder
2. You open a random file from another project (outside your workspace)
3. You have many projects in one folder which depend on each other (ala crates)